File: crates/engine/src/seams/refresh_hint.rs
//! `RefreshHintSource` — host events forcing an immediate tick
//! (blueprint/engine.md).

/// One refresh hint. Carries no payload in v2.0: a hint only ever means
/// "run an immediate focus-window tick now".
///
/// The designed-for push overlay (#33 D1 — API WebSocket hints, desktop
/// PubSub) extends this type behind the same seam when it lands; hosts and
/// the engine already treat hints as events, so richer payloads are
/// additive.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct RefreshHint;

/// Host-event stream that forces an immediate sync tick.
///
/// Web: UI navigation, tab-visibility regain, `online` reconnect, cross-tab
/// focus hints. Desktop: FUSE-op TTL checks, reconnect. Hints are
/// best-effort accelerators — losing one costs staleness bounded by the
/// poll cadence, never correctness — so no conformance kit ships for this
/// seam: it has no durable semantics, and its events are host-driven.
pub trait RefreshHintSource {
    /// Waits for the next hint; `None` once the source is closed for good
    /// (host shutdown). The engine treats `None` as end-of-stream and stops
    /// listening.
    async fn next_hint(&mut self) -> Option<RefreshHint>;
}

File: packages/client/src/seams/refreshHint.ts
/**
 * `RefreshHintSource` — a queue of host events that force an immediate sync tick
 * (blueprint/web-client.md seam table).
 *
 * The host pushes a hint on UI navigation, `visibilitychange` regain, `online`
 * reconnect, or a cross-tab focus signal; the engine awaits `nextHint`. Hints
 * carry no payload (v2.0) and are best-effort accelerators — a dropped hint
 * costs staleness bounded by the poll cadence, never correctness.
 */

import type { RefreshHintSourceSeam } from './types.js';

export class QueueRefreshHintSource implements RefreshHintSourceSeam {
  private pending = 0;
  private waiters: Array<(hint: true | null) => void> = [];
  private closed = false;

  /** Signals one refresh hint (coalesced only by the engine, never dropped here). */
  pushHint(): void {
    if (this.closed) return;
    const waiter = this.waiters.shift();
    if (waiter) waiter(true);
    else this.pending += 1;
  }

  /** Closes the source for good; the engine stops listening. */
  close(): void {
    if (this.closed) return;
    this.closed = true;
    const waiters = this.waiters;
    this.waiters = [];
    for (const waiter of waiters) waiter(null);
  }

  nextHint(): Promise<true | null> {
    if (this.pending > 0) {
      this.pending -= 1;
      return Promise.resolve(true);
    }
    if (this.closed) return Promise.resolve(null);
    return new Promise((resolve) => this.waiters.push(resolve));
  }
}
